core_v14/src/decoder/mod.rs

```diff
@@ -155,32 +154,43 @@ impl Decoder {
 	/// expected to be provided in a SCALE-encoded form equivalent to `Vec<(Compact<u32>,Extrinsic)>`; in other words, we
 	/// start with a compact encoded count of how many extrinsics exist, and then each extrinsic is prefixed by
 	/// a compact encoding of its byte length.
-	pub fn decode_extrinsics(&self, data: &[u8]) -> Result<Vec<Extrinsic>, (Vec<Extrinsic>, DecodeError)> {
-		let extrinsic_bytes = AllExtrinsicBytes::new(data).map_err(|e| (Vec::new(), e.into()))?;
+	pub fn decode_extrinsics<'a>(
+		&'a self,
+		data: &mut &[u8],
+	) -> Result<Vec<Extrinsic<'a>>, (Vec<Extrinsic<'a>>, DecodeError)> {
+		let extrinsic_bytes = AllExtrinsicBytes::new(*data).map_err(|e| (Vec::new(), e.into()))?;
 
 		log::trace!("Decoding {} Total Extrinsics.", extrinsic_bytes.len());
 
 		let mut out = Vec::with_capacity(extrinsic_bytes.len());
-		for (idx, res) in extrinsic_bytes.iter().enumerate() {
+		let mut extrinsics_iter = extrinsic_bytes.iter();
+		for res in &mut extrinsics_iter {
 			let single_extrinsic = match res {
 				Ok(bytes) => bytes,
 				Err(e) => return Err((out, e.into())),
 			};
 
-			log::trace!("Extrinsic {}:{:?}", idx, single_extrinsic.bytes());
+			log::trace!("Extrinsic:{:?}", single_extrinsic.bytes());
 
-			let ext = match self.decode_unwrapped_extrinsic(single_extrinsic.bytes()) {
+			let bytes = &mut single_extrinsic.bytes();
+			let ext = match self.decode_unwrapped_extrinsic(bytes) {
 				Ok(ext) => ext,
-				Err(DecodeError::LateEof(remaining, ext)) => {
-					// Returned from `decode_extrinsics`, we want the remaining bytes to be relative
-					// to the data slice passed in, and not the single extrinsic slice.
-					return Err((out, DecodeError::LateEof(single_extrinsic.remaining() + remaining, ext)));
-				}
 				Err(e) => return Err((out, e)),
 			};
 
+			// If decoding didn't consume all extrinsic bytes, something went wrong.
+			// Hand back whatever we have but note the error.
+			if !bytes.is_empty() {
+				return Err((out, DecodeError::DidntConsumeBytes(bytes.len())));
+			}
+
 			out.push(ext);
 		}
+
+		// Shift our externally provided data cursor forwards to the right spot,
+		// so that one can continue to decode more bytes if there are any:
+		*data = extrinsics_iter.remaining_bytes();
```

Comment:
There isn't an obvious case from what we'd be decoding I think. This is more just so that all of the `decode_x` calls are consistent in the way that they will leave remaining bytes over instead of erroring.